### DISCUSSION TITLE
Where to Implement `AppLabel`

### DISCUSSION THREAD

**Discussion by zicklag** [bevyengine/bevy#2629] [Q&A] [answered]
I was going to work on using labels instead of indices for sub-apps for the new pipelined renderer, but I wanted some guidance on where we wanted the label code implemented. Currently all of our labels are trapped in the `bevy_ecs` crate, and their derive macros are in `bevy_ecs_macros`, but if we need an `AppLabel` type, then it kind of makes most sense to put it in the `bevy_app` crate.

In that case I think we would need a new `bevy_app_macros` crate, and we would end up copying the label creation and derive code from the `bevy_ecs` crate. Is that the direction we want to take? It seems like it would be nice not to have to duplicate the code, but it seems weird to shove the `AppLabel` inside of `bevy_ecs` where it isn't used.

If we wanted to avoid some extra code duplication we could create a new `bevy_label` crate that has the shared code for the traits and macros, but I don't know if we want to add yet another crate just for that.

**Comment by TheRawMeatball** [comment, marked as answer]
I think putting the majority of the code for implementing labels in a separate crate, or perhaps bevy_utils would be the correct choice

**Reply by cart** [reply]
Yeah I think generic label code should probably live in bevy_utils for now.
